docs/hotfix-deployment.md

```diff
@@ -0,0 +1,113 @@
+# Apply for Postgraduate Teacher Training - Hotfix Deployment
+
+## Purpose
+
+This document describes the process of manually deploying a hotfix
+
+### When should this process be used?
+
+Normally deployments to the `staging`, `sandbox` and `production` environments
+happen on a daily (or more frequent) basis and include all changes that
+have been merged into master. A hotfix is a more urgent deployment
+containing typically just one pull request. A hotfix should be performed
+when:
+
+1. The fix is urgent.
+2. There are other changes already merged into master since the last
+   deployment that we don't want to deploy yet perhaps because they are
+   not fully tested.
+
+## Instructions
+
+We assume that we begin in a situation like this:
+
+```
+master        ---D--------x-----x--------------x--->
+                  \      /     / \            /
+feature1           -x-x-x     /   \          /
+                   \         /     \        /
+feature2            --x-x-x-x       \      /
+                                     \    /
+feature3                              -x-x
+```
+
+Since last deployed at `D` we've merged two additional features into
+`master`. We urgently need to create a fix that does not include
+features 1, 2 and 3, we just want everything up to `D`.
+
+### Minimum steps that would work...
+
+1. Fetch `master` to your local machine and create a branch from `D`
+   (not `HEAD` as we normally would).
+2. Implement the fix locally raise a PR and get it approved in the
+   normal way. Test it locally or using the Heroku review app that will
+   be automatically created.
+3. Deploy the `hotfix` branch using the normal deployment procedure
+   (except picking `HEAD` of the `hotfix` branch rather than a specific
+   SHA on `master` as we normally do).
+4. Merge the `hotfix` branch back to `master`.
+
+```
+hotfix             ------------------------------x-x-x-H
+                  /                                     \
+master        ---D--------x-----x--------------x----------->
+                  \      /     / \            /
+feature1           -x-x-x     /   \          /
+                   \         /     \        /
+feature2            --x-x-x-x       \      /
+                                     \    /
+feature3                              -x-x
+```
+
+### Some questions
+
+Do we need to tag releases? At the moment we don't really have record of
+which versions shipped and when in the git repository itself. It makes
+sense to tag releases that ship to production so that we have a readily
+available record of that. It would make finding the commit from which a
+hotfix should be branched a little easier.
+
+Should we get into the habit of creating a release branch for normal
+releases? (A release branch is really a generalisation of a hotfix
+branch)
+
+How can we automate step 4 (merging the hotfix back to master)? We want
```

Comment:
Hotfixes are meant to be rare events, so asking the hotfixer to raise a PR for merging hotfix into master seems fine to me. We could even leverage the need to create a PR as a confirmation step that the hotfix has worked. Finally, we should probably extend the apply ops dashboard to detect if a hotfix is in progress.